src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetExcelGenerator.java

```diff
@@ -164,27 +164,39 @@ public File generateTraitAndSelectionVariablesFile(final List<String[]> rowValue
 		throw new UnsupportedOperationException();
 	}
 
-	List<MeasurementVariable> orderColumns(final List<MeasurementVariable> columns) {
+	List<MeasurementVariable> orderColumns(final List<MeasurementVariable> columns, final int firstColumnVariableId) {
 		final List<MeasurementVariable> orderedColumns = new ArrayList<>();
 		final List<MeasurementVariable> trait = new ArrayList<>();
 		final List<MeasurementVariable> selection = new ArrayList<>();
+		final List<MeasurementVariable> genotypeMarkers = new ArrayList<>();
 
+		MeasurementVariable observationUnitVariable = null;
 		for (final MeasurementVariable measurementVariable : columns) {
-			if (TermId.OBS_UNIT_ID.getId() == measurementVariable.getTermId()) {
+			if (measurementVariable.getVariableType() != null && VariableType.OBSERVATION_UNIT.getId()
+					.equals(measurementVariable.getVariableType().getId())) {
+				observationUnitVariable = measurementVariable;
+			} else if (firstColumnVariableId == measurementVariable.getTermId()) {
```

Comment:
Move this first variable logic outside of for loop and add comment why it is necessary.